### ISSUE TITLE
Error opening a list of blob containers in Docker container hosted emulator

### ISSUE THREAD

**Issue by longtimeago** [microsoft/AzureStorageExplorer#3564]
**Storage Explorer Version**: 1.15.1
**Build Number**: 20200902.9
**Platform/OS**: Linux Ubuntu 20.04
**Architecture**: x64
**Regression From**:  Didn't try

## Bug Description ##
Error opening a list of Blob Containers

## Steps to Reproduce ##
1. Connect via connection string to a local azure store (azurite)
`DefaultEndpointsProtocol=http;AccountName=devstoreaccount1;AccountKey=xxx;BlobEndpoint=http://azurite:10000/devstoreaccount1;`
2. Open Blob Containers in Explorer tree
3. Error pops up
4. If I click `Restart` the error pops up on every restart
5. I can press `Ignore` then containers are listed

## Expected Experience ##
Shouldn't be an error

## Actual Experience ##
![image](https://user-images.githubusercontent.com/1388994/92877922-4b6a4380-f414-11ea-97a4-6329e36e512d.png)

## Additional Context ##
Azurite v2.6.5, running in Docker

**Comment by craxal** [comment]
@longtimeago I notice your emulator endpoint is http://azurite:10000, which isn't actually a local endpoint. Have you tried http://localhost:10000? Also, we have a node in the tree view that connects to the default local endpoints automatically. Do those nodes work with your emulator?

**Comment by longtimeago** [comment]
> I notice your emulator endpoint is http://azurite:10000, which isn't actually a local endpoint.

@craxal well, yea, it's _local_ comparing to a _cloud_ version but in fact is running in a docker container on a local machine.
The proper connection string was 
`DefaultEndpointsProtocol=http;AccountName=devstoreaccount1;AccountKey=xxx;BlobEndpoint=http://172.18.0.30:10000/devstoreaccount1;` where `172.18.0.30` is a container address with azurite simulator running.

> Have you tried http://localhost:10000

yes, when I run azurite on via npm (without docker) and connect to it via localhost there is no such error

> Also, we have a node in the tree view that connects to the default local endpoints automatically. Do those nodes work with your emulator?

I you mean node `local (Key)` under `Storage Accounts` than it's present but empty

**Comment by craxal** [comment]
@longtimeago Ah, yeah, then none of the default stuff is going to work. Sorry, I missed the Docker context detail you mentioned. Running an emulator from a Docker container is not a scenario we've designed or tested for. Two questions for you:
- Does it work if the emulator is truly running on locahost?
- Are you running the emulator in a container to meet certain requirements? In other words, what are your reasons for needing or wanting to run the emulator in a container?

**Comment by longtimeago** [comment]
> * Does it work if the emulator is truly running on locahost (this is to make sure you have some viable workaround)?

yes, it works :+1: 

About motivation for running the emulator in a container
1) We use it in integration tests; the app we are working on depends on Azure (as well as Amazon S3 and Google) blob stores. Thus, it's very convenient to run entire stack including Azure emulator in separate containers
2) From (backend) developer's POV it's cool to have portable environment with minimum prerequisites to start the stack locally - emulator shouldn't be installed locally (requires npm by itself) but merely running a docker container with emulator.

And of course in both cases it's quite important sometimes to login to a the emulator and check blobs by eyes :)

@craxal Hope that explains our motivation pretty much :)

**Comment by craxal** [comment]
@longtimeago Thank you, yes, that is very helpful.

Looks like #3153 could be related. We will look into scheduling some time to investigate this further.

**Comment by longtimeago** [comment]
@craxal sweet, thanks a lot!

**Comment by craxal** [comment]
@longtimeago So, I just verified that Storage Explorer can, in fact, connect to an emulator hosted on a non-local machine on the same local network using the following connection string:
```txt
DefaultEndpointsProtocol=http;
AccountName=devstoreaccount1;
AccountKey=Eby8vdM02xNOcqFlqUwJPLlmEtlCDXJ1OUzFT50uSRZ6IFsuFq2UVErCz4I6tq/K1SZFPTOtr/KBHBeksoGMGw==;
BlobEndpoint=http://[EMULATOR HOST IP ADDRESS]:10000/devstoreaccount1;
QueueEndpoint=http://[EMULATOR HOST IP ADDRESS]:10001/devstoreaccount1;
```

Next, I'll be setting up a Docker image. I'm assuming you've been using the [documented Azurite Docker image](https://docs.microsoft.com/en-us/azure/storage/common/storage-use-azurite#install-and-run-the-azurite-docker-image)? Did you use a different command to start the container?

**Comment by craxal** [comment]
@longtimeago I've now just verified that Storage Explorer can connect to an emulator running in a local Docker container. I didn't have to use a special connection string at all when Azurite was configured with the default ports. The "(Emulator - Default Ports)" node successfully connected.

When starting the container, did you do this command?
```
docker run -p 10000:10000 -p 10001:10001 mcr.microsoft.com/azure-storage/azurite
```

**Comment by longtimeago** [comment]
> Next, I'll be setting up a Docker image. I'm assuming you've been using the [documented Azurite Docker image](https://docs.microsoft.com/en-us/azure/storage/common/storage-use-azurite#install-and-run-the-azurite-docker-image)?

yes

> docker run -p 10000:10000 -p 10001:10001 mcr.microsoft.com/azure-storage/azurite

@craxal we use docker-compose but that shouldn't matter.

IMO, the main reason it works for you is because of exposing ports to a localhost `-p 10000:10000 -p 10001:10001`
We don't do that because that would prevent running more than one azurite container on the same host (which we need)

Try to remove `-p 10000:10000 -p 10001:10001` from your command and use the same connection string as in your previous comment. You should see the error :)

**Comment by longtimeago** [comment]
> Also, out of curiosity, do you have any Storage Explorer proxy settings enabled?

noup, just stock config, nothing fancy :)

**Comment by craxal** [comment]
@longtimeago Yes, that would cause an error, but I would expect that; you have to connect the ports up. I'm not really sure how you can send requests to an emulator inside of a container without connecting the container ports to the local host.

Using multiple emulators on one machine should be easy, though. You just have to connect different ports up, so the host knows to which emulator to send requests. For example:

```bash
# Forward requests sent to localhost:10000 to port 10000 on the first container.
docker run -p 10000:10000 mcr.microsoft.com/azure-storage/azurite

# Forward requests sent to localhost:20000 to port 10000 on the second container 
docker run -p 20000:10000 mcr.microsoft.com/azure-storage/azurite
```

**Comment by longtimeago** [comment]
> ```shell
> # Forward requests sent to localhost:10000 to port 10000 on the first container.
> ```

@craxal there are 2 downsides of this approach
1) It pollutes localhost with occupied ports, which is definitely a bad thing if I run many other containers
2) Instead of tracking just ip addresses per docker container we have to track IP address AND ports

> I'm not really sure how you can send requests to an emulator inside of a container without connecting the container ports to the local host.

@craxal every docker container has IP address in docker network. This network is bridged to localhost. It means knowing the container IP I can access ports of services inside the container.

Container IP address can be found as `docker inspect -f '{{range.NetworkSettings.Networks}}{{.IPAddress}}{{end}}' container_name_or_id`

Having that you should be able to access the service inside a container using found IP address and default, not exposed/forwarded, port - 10000 in case of azurite BlobEndpoint

Just checked with standalone container running without exposed ports 
1. ```docker run mcr.microsoft.com/azure-storage/azurite```
2. Found container IP with command above
3. Tried to connect to the port using `netcat IP_ADDRESS 10000`

Oh, and I'm on Ubuntu
If you are on Win or Mac this way of container access may not work because on those platforms docker daemon itself is running on a mini VM, so may need to bridge docker network inside that VM to your host.

**Comment by craxal** [comment]
@longtimeago Sorry for the long delay. In the steps you mentioned previously, you said you "tried to connect". Were you successful? From where did you try to connect?

Let's see if I understand the scenario you have. You are hosting multiple Docker containers on one Ubuntu machine. Each Docker container contains an instance of the Azurite emulator. It is your intention to allow a network of machines access to these emulators. You are concerned about consuming lots of ports on the Docker host machine, so you want let users use an individual IP address to access a container.

I'd like to try and set something up like that and test Storage Explorer with it, but I could use your help in knowing how. I'm not finding much documentation for setting up containers with no port mapping.

**Comment by longtimeago** [comment]
> you said you "tried to connect". Were you successful? From where did you try to connect?

@craxal Yes, netcat connection succeed. From command line on my host

> Let's see if I understand the scenario you have.

I think you are complicating things. This is how to reproduce the issue (on Ubuntu host)
1. I run docker emulator in a container (with `docker run mcr.microsoft.com/azure-storage/azurite`)
2. Find `CONTAINER_IP` address `docker inspect -f '{{range.NetworkSettings.Networks}}{{.IPAddress}}{{end}}' container_name_or_id`
3. Open storage explorer and use local connection `DefaultEndpointsProtocol=http;AccountName=devstoreaccount1;AccountKey=xxx;BlobEndpoint=http://CONTAINER_IP:10000/devstoreaccount1;` 

I expect it to work, but it does not with an error in the task description

This is the simplest way of reproduction of the issue :)

If you are really curious about our use-case than here it is
- we run several docker containers with docker-compose
- some of those containers consumes Azure services (push blobs)
- to simulate real Azure services in integration tests we added one more docker container with Azure emulator to docker-compose
- sometimes we want to check what the emulator has inside after test failure
- for that we have storage explorer installed on host machine and trying to connect to Azure emulator

**Comment by craxal** [comment]
@longtimeago Sorry again for the long delay! We haven't forgotten about your issue.

I was successful at connecting to a local emulator hosted in a Docker container. I could list containers, create new containers, and list blobs in the containers. The Docker container output confirms the emulator is accepting requests. No port forwarding necessary.

I'm running Storage Explorer 1.21.3. We've made some significant changes since you first opened you're issue, so it's possible we may have fixed the issue without specifically looking at it.

Here are the steps I followed:
1. Start the emulator in a container:

   ```
   sudo docker run mcr.microsoft.com/azure-storage/azurite
   ```

2. Get the IP address of the container:

   ```
   docker inspect -f '{{range.NetworkSettings.Networks}}{{.IPAddress}}{{end}}' container_name_or_id
   ```

3. Launch Storage Explorer
4. Click the plug icon in the vertical toolbar, and select "Storage account or service" (_not_ "Local storage emulator").
5. Select "Connection string (Key or SAS)" and click Next.
6. Paste the connection string and finish creating the connection:
   ```
   DefaultEndpointsProtocol=http;
   AccountName=devstoreaccount1;
   AccountKey=[EMULATOR KEY];
   BlobEndpoint=http://[EMULATOR HOST IP ADDRESS]:10000/devstoreaccount1;
   QueueEndpoint=http://[EMULATOR HOST IP ADDRESS]:10001/devstoreaccount1;
   ```

If there's anything you're doing differently, let me know, and we can keep digging.
